src/multi.c

```diff
@@ -379,14 +403,31 @@ void touchAllWatchedKeysInDb(redisDb *emptied, redisDb *replaced_with) {
     dictIterator *di = dictGetSafeIterator(emptied->watched_keys);
     while((de = dictNext(di)) != NULL) {
         robj *key = dictGetKey(de);
-        if (dictFind(emptied->dict, key->ptr) ||
+        int exists_in_emptied = dictFind(emptied->dict, key->ptr) != NULL;
+        if (exists_in_emptied ||
             (replaced_with && dictFind(replaced_with->dict, key->ptr)))
         {
             list *clients = dictGetVal(de);
             if (!clients) continue;
             listRewind(clients,&li);
             while((ln = listNext(&li))) {
-                client *c = listNodeValue(ln);
+                watchedKey *wk = listNodeValue(ln);
+                if (wk->expired) {
+                    if (!replaced_with || !dictFind(replaced_with->dict, key->ptr)) {
+                        /* Expired key now deleted. No logical change. Clear the
+                         * flag. Deleted keys are not flagged as expired. */
+                        wk->expired = 0;
+                        continue;
+                    } else if (keyIsExpired(replaced_with, key)) {
+                        /* Expired key remains expired. */
+                        continue;
+                    }
```

Comment:
maybe leave a comment here is better:
```c
// expired flag should be set to 0 when the expired key replaced with a non-expired key, but we will call unwatchAllKeys() after the scope, so it's safe to skip it.
// else {
//     wk->expired = 0;
// }
```

but it's not a big deal, you can ignore my words.

Reply:
We do `continue` in a lot of exceptional cases, but the main flow is the unwatchAllKeys below. Maybe it looks better if we do unwatchAllKeys in an else-branch and remove the continue in all these cases. WDYT?

```C
                if (wk->expired &&
                    (!replaced_with || !dictFind(replaced_with->dict, key->ptr)))
                {
                    /* Expired key now deleted. No logical change. Clear the
                     * flag. Deleted keys are not flagged as expired. */
                    wk->expired = 0;
                } else if (wk->expired && keyIsExpired(replaced_with, key)) {
                    /* Expired key remains expired. */
                } else if (!exists_in_emptied && keyIsExpired(replaced_with, key)) {
                    /* Non-existing key is replaced with an expired key. */
                    wk->expired = 1;
                } else {
                    client *c = wk->client;
                    c->flags |= CLIENT_DIRTY_CAS;
                    /* As the client is marked as dirty, there is no point in
                     * getting here again for others keys (or keep the memory
                     * overhead till EXEC). */
                    unwatchAllKeys(c);
                }
```

Reply:
i think the unwatch is the common case, and all the others are special cases.. so the `continue` approach is preferred IMHO, but i don't care much either way.

Reply:
OK, I'm not changing anything then.